File: systems-for-design-and-implementation/labs/exam/TicTacToePersistance/src/tictactoe/persistance/repository/jdbc/DBConnection.java
package tictactoe.persistance.repository.jdbc;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.File;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

public class DBConnection
{
    private static DBConnection instance;
    private static String XML_FILE = "database.xml";
    private Connection connection;

    private DBConnection()
    {
        initConnection();
    }

    public static DBConnection get()
    {
        if (instance == null)
        {
            instance = new DBConnection();
        }

        return instance;
    }

    private void initConnection()
    {
        try
        {
            File fXmlFile = new File(XML_FILE);
            DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
            DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
            Document doc = dBuilder.parse(fXmlFile);

            //optional, but recommended
            //read this - http://stackoverflow.com/questions/13786607/normalization-in-dom-parsing-with-java-how-does-it-work
            doc.getDocumentElement().normalize();

            Element configElement = doc.getDocumentElement();

            String driver = configElement.getElementsByTagName("driver").item(0).getTextContent();
            String username = configElement.getElementsByTagName("username").item(0).getTextContent();
            String password = configElement.getElementsByTagName("password").item(0).getTextContent();
            String database = configElement.getElementsByTagName("database").item(0).getTextContent();
            String url = configElement.getElementsByTagName("url").item(0).getTextContent() + database;

            try
            {
                Class.forName(driver);
                System.out.println(String.format("Driver %s loaded", driver));

                connection = DriverManager.getConnection(url, username, password);
            }
            catch (ClassNotFoundException e)
            {
                System.out.println("Error loading driver " + e);
            }
            catch (SQLException e)
            {
                System.out.println("Error getting connection " + e);
            }

        }
        catch (Exception e)
        {
            e.printStackTrace();
        }

        if (connection == null)
        {
            throw new RuntimeException("The SQL Connection was not set");
        }
    }

    public Connection getConnection()
    {
        try
        {
            if (connection != null && connection.isClosed())
            {
                initConnection();
            }
        }
        catch (SQLException e)
        {
            System.out.println("Error DB " + e);
        }

        return connection;
    }
}

File: systems-for-design-and-implementation/labs/lab3/LibraryPersistance/src/library/persistance/repository/jdbc/DBConnection.scala
package library.persistance.repository.jdbc

import java.io.File
import java.sql.{Connection, DriverManager, SQLException}
import javax.xml.parsers.{DocumentBuilder, DocumentBuilderFactory}

import org.w3c.dom.{Document, Element}

object DBConnection
{
    private val XML_FILE: String = "database.xml"
    private var instance: DBConnection = null

    def get: DBConnection =
    {
        if (instance == null)
        {
            instance = new DBConnection
        }
        instance
    }
}

class DBConnection private
{
    private var connection: Connection = null
    initConnection()

    def getConnection: Connection =
    {
        try
        {
            if (connection != null && connection.isClosed)
            {
                initConnection()
            }
        }
        catch
        {
            case e: SQLException =>
                println("Error DB " + e)
        }
        connection
    }

    private def initConnection()
    {
        try
        {
            val fXmlFile: File = new File(DBConnection.XML_FILE)
            val dbFactory: DocumentBuilderFactory = DocumentBuilderFactory.newInstance
            val dBuilder: DocumentBuilder = dbFactory.newDocumentBuilder
            val doc: Document = dBuilder.parse(fXmlFile)
            doc.getDocumentElement.normalize()
            val configElement: Element = doc.getDocumentElement
            val driver: String = configElement.getElementsByTagName("driver").item(0).getTextContent
            val username: String = configElement.getElementsByTagName("username").item(0).getTextContent
            val password: String = configElement.getElementsByTagName("password").item(0).getTextContent
            val database: String = configElement.getElementsByTagName("database").item(0).getTextContent
            val url: String = configElement.getElementsByTagName("url").item(0).getTextContent + database
            try
            {
                Class.forName(driver)
                println(String.format("Driver %s loaded", driver))
                connection = DriverManager.getConnection(url, username, password)
            }
            catch
            {
                case e: ClassNotFoundException =>
                    println("Error loading driver " + e)
                case e: SQLException =>
                    println("Error getting connection " + e)
            }
        }
        catch
        {
            case e: Exception => e.printStackTrace()
        }
        if (connection == null)
        {
            throw new RuntimeException("The SQL Connection was not set")
        }
    }
}
